ChopChop/ViewModels/Recipes/RecipeFormViewModel.swift

```
@@ -0,0 +1,237 @@
+//  ChopChop
+//
+//  Created by Cao Wenjie on 20/3/21.
+//
+import SwiftUI
+import Combine
+import GRDB
+
+class RecipeFormViewModel: ObservableObject {
+
+    private var existingRecipe: Recipe?
+    private var recipeId: Int64?
+    private let storageManager = StorageManager()
+    private var recipeCategoryCancellable = Set<AnyCancellable>()
+
+    private(set) var errorMessage = ""
+    private(set) var isEdit = false
+
+    var pickerSourceType: UIImagePickerController.SourceType = .photoLibrary
+
+    @Published var hasError: Bool = false
+    @Published var isShowingPhotoLibrary = false
+    @Published var image = UIImage()
+    @Published var recipeName = ""
+    @Published var serving = "" {
+        didSet {
+            ensureValidServing()
+        }
+    }
+    @Published var allRecipeCategories = [RecipeCategory]()
+    @Published var recipeCategory = ""
+    @Published var difficulty: Difficulty = .veryEasy
+    @Published var steps = [String]()
+    @Published var ingredients = [RecipeIngredientRowViewModel]()
+    @Published var ingredientParsingString = ""
+    @Published var instructionParsingString = ""
+
+    init(recipe: Recipe) {
+        existingRecipe = recipe
+        recipeId = recipe.id
+        recipeName = recipe.name
+        serving = recipe.servings.description
+        difficulty = recipe.difficulty ?? .veryEasy
+
+        steps = recipe.steps.map({ $0.content })
+        ingredients = recipe.ingredients.map({
+            RecipeIngredientRowViewModel(
+                amount: $0.quantity.value.description,
+                unit: $0.quantity.type,
+                ingredientName: $0.name
+            )
+        })
+        image = storageManager.fetchRecipeImage(name: recipe.name) ?? UIImage()
+        fetchCategories()
+        if let categoryId = recipe.recipeCategoryId {
+            do {
+                recipeCategory = try storageManager.fetchRecipeCategory(id: categoryId)?.name ?? ""
+            } catch {
```

possible to rm empty catch block? can use guard conditions with try?